sklearn/ensemble/_forest.py

```diff
@@ -398,13 +408,63 @@ def fit(self, X, y, sample_weight=None):
         if self.oob_score:
             self._set_oob_score(X, y)
 
+        if self.feature_importances_type == "permutation":
+            self._set_oob_permutation_importance(X, y)
+
         # Decapsulate classes_ attributes
         if hasattr(self, "classes_") and self.n_outputs_ == 1:
             self.n_classes_ = self.n_classes_[0]
             self.classes_ = self.classes_[0]
 
         return self
 
+    def _set_oob_permutation_importance(self, X, y):
+        """
+        Calculate out of bag predictions and score."""
+        X = check_array(X, dtype=DTYPE, accept_sparse=False)
+
+        n_samples = y.shape[0]
+
+        n_predictions = np.zeros(n_samples)
+
+        n_samples_bootstrap = _get_n_samples_bootstrap(
+            n_samples, self.max_samples
+        )
+
+        all_imp = np.zeros((self.n_estimators, X.shape[1]))
+        for i, estimator in enumerate(self.estimators_):
+            random_state = check_random_state(estimator.random_state)
+
+            unsampled_indices = _generate_unsampled_indices(
+                estimator.random_state, n_samples, n_samples_bootstrap
+            )
+
+            scorer = check_scoring(estimator, scoring=self.scoring)
+            baseline = scorer(
+                estimator, X[unsampled_indices, :], y[unsampled_indices, :]
+            )
+
+            for col in range(X.shape[1]):
+                X_permuted = X.copy()
+                random_state.shuffle(X_permuted[:, col])
+                curr_perf = scorer(
```

Comment:
Actually, I am having a second thought here. I give an additional read to Breiman's paper after looking at our function that computes the OOB score.

In your proposal, we average the score of each individual tree while I think that the idea would be to aggregate the votes of each tree and finally (I assume with majority vote / mean) compare with the true target. This is indeed what is intended in the OOB scoring function.

I am not sure that both approaches are equivalent. In this way it seems that we should have a function that should build the OOB predictions that could be reuse between the OOB score and the permutation OOB.

@NicolasHug @ogrisel Could you help regarding the theoretical part.

Then I assume that we can fusion the OOB scoring where we can compute the permutation score if required instead of doing the same manipulation twice.

Reply:
I considered that approach when writing this PR, and it might be what Breiman is suggesting, although he doesn't go into much detail. To be clear, these are the two options:

1. Get OOB performance degradation for each tree individually when a feature is permuted, then average those degradations across all trees (proposed in this PR)
2. Get OOB performance degradation across the whole RF when a feature is permuted

I agree with you that the 2nd sounded better to me at first, but I chose the first for 2 reasons:
* It was much more stable in the tests I ran (i.e. importances didn't change as much due to randomness)
* It's the same process used in the R `Ranger` package, as detailed [here](https://link.springer.com/article/10.1186/s12859-016-0995-8):
> The basic idea of the permutation variable importance approach [18] is to consider a variable important if it has a positive effect on the prediction performance. To evaluate this, a tree is grown in the first step, and the prediction accuracy in the OOB observations is calculated. In the second step, any association between the variable of interest x i and the outcome is broken by permuting the values of all individuals for x i , and the prediction accuracy is computed again. The difference between the two accuracy values is the permutation importance for x i from a single tree. The average of all tree importance values in a random forest then gives the random forest permutation importance of this variable. The procedure is repeated for all variables of interest. The package ranger [24] was used in our analyses.

I'm happy to change it to the second option if that's preferred. My vote would be for the first though